File: tools/tf-types-ts/src/core/envelope.ts
import type { SignatureEnvelope } from "../generated/_common.js";

export type EnvelopeIssue =
  | { code: "missing-algorithm" }
  | { code: "missing-signer" }
  | { code: "missing-signature" }
  | { code: "invalid-base64"; field: "signature" | "alt_signature" }
  | { code: "alt-without-algorithm" }
  | { code: "unknown-algorithm"; algorithm: string }
  | { code: "unknown-alt-algorithm"; algorithm: string };

const KNOWN_ALGORITHMS: ReadonlySet<string> = new Set([
  "ed25519",
  "ed448",
  "p256",
  "p384",
  "p521",
  "rsa-pss-sha256",
  "ml-dsa-44",
  "ml-dsa-65",
  "ml-dsa-87",
  "slh-dsa-sha2-128s",
  "slh-dsa-sha2-192s",
]);

const BASE64_RE = /^[A-Za-z0-9+/]+={0,2}$/;

export interface ValidationResult {
  readonly ok: boolean;
  readonly issues: EnvelopeIssue[];
}

/**
 * Validate the shape of a SignatureEnvelope. Does NOT verify the signature;
 * no crypto is performed in the foundation phase. Unknown algorithms
 * produce non-fatal warnings in the issue list but still pass `ok=true`.
 */
export function validateEnvelopeShape(e: SignatureEnvelope): ValidationResult {
  const issues: EnvelopeIssue[] = [];
  if (!e.algorithm) issues.push({ code: "missing-algorithm" });
  if (!e.signer) issues.push({ code: "missing-signer" });
  if (!e.signature) issues.push({ code: "missing-signature" });
  if (e.signature && !BASE64_RE.test(e.signature)) issues.push({ code: "invalid-base64", field: "signature" });
  if (e.alt_signature && !BASE64_RE.test(e.alt_signature)) issues.push({ code: "invalid-base64", field: "alt_signature" });
  if (e.alt_signature && !e.alt_algorithm) issues.push({ code: "alt-without-algorithm" });

  const fatalCount = issues.filter((i) => !isWarning(i)).length;
  if (e.algorithm && !KNOWN_ALGORITHMS.has(e.algorithm)) {
    issues.push({ code: "unknown-algorithm", algorithm: e.algorithm });
  }
  if (e.alt_algorithm && !KNOWN_ALGORITHMS.has(e.alt_algorithm)) {
    issues.push({ code: "unknown-alt-algorithm", algorithm: e.alt_algorithm });
  }
  return { ok: fatalCount === 0, issues };
}

function isWarning(i: EnvelopeIssue): boolean {
  return i.code === "unknown-algorithm" || i.code === "unknown-alt-algorithm";
}

File: crates/tf-types/src/envelope.rs
//! Signature envelope shape validator — mirrors
//! `tools/tf-types-ts/src/core/envelope.ts`. No crypto is performed here;
//! real signing/verification lives in `crypto.rs`.

use crate::generated::common::SignatureEnvelope;

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum EnvelopeIssue {
    MissingAlgorithm,
    MissingSigner,
    MissingSignature,
    InvalidBase64 { field: &'static str },
    AltWithoutAlgorithm,
    UnknownAlgorithm { algorithm: String },
    UnknownAltAlgorithm { algorithm: String },
}

impl EnvelopeIssue {
    pub fn is_warning(&self) -> bool {
        matches!(
            self,
            EnvelopeIssue::UnknownAlgorithm { .. } | EnvelopeIssue::UnknownAltAlgorithm { .. }
        )
    }
}

pub struct EnvelopeValidation {
    pub ok: bool,
    pub issues: Vec<EnvelopeIssue>,
}

const KNOWN_ALGORITHMS: &[&str] = &[
    "ed25519",
    "ed448",
    "p256",
    "p384",
    "p521",
    "rsa-pss-sha256",
    "ml-dsa-44",
    "ml-dsa-65",
    "ml-dsa-87",
    "slh-dsa-sha2-128s",
    "slh-dsa-sha2-192s",
];

pub fn validate_envelope_shape(e: &SignatureEnvelope) -> EnvelopeValidation {
    let mut issues = Vec::new();

    if e.algorithm.is_empty() {
        issues.push(EnvelopeIssue::MissingAlgorithm);
    }
    if e.signer.is_empty() {
        issues.push(EnvelopeIssue::MissingSigner);
    }
    if e.signature.is_empty() {
        issues.push(EnvelopeIssue::MissingSignature);
    }
    if !e.signature.is_empty() && !is_base64(&e.signature) {
        issues.push(EnvelopeIssue::InvalidBase64 { field: "signature" });
    }
    if let Some(alt) = &e.alt_signature {
        if !is_base64(alt) {
            issues.push(EnvelopeIssue::InvalidBase64 {
                field: "alt_signature",
            });
        }
        if e.alt_algorithm.is_none() {
            issues.push(EnvelopeIssue::AltWithoutAlgorithm);
        }
    }

    let fatal_count = issues.iter().filter(|i| !i.is_warning()).count();

    if !e.algorithm.is_empty() && !KNOWN_ALGORITHMS.contains(&e.algorithm.as_str()) {
        issues.push(EnvelopeIssue::UnknownAlgorithm {
            algorithm: e.algorithm.clone(),
        });
    }
    if let Some(alt) = &e.alt_algorithm {
        if !KNOWN_ALGORITHMS.contains(&alt.as_str()) {
            issues.push(EnvelopeIssue::UnknownAltAlgorithm {
                algorithm: alt.clone(),
            });
        }
    }

    EnvelopeValidation {
        ok: fatal_count == 0,
        issues,
    }
}

fn is_base64(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    let mut eq_seen = 0;
    for c in s.chars() {
        match c {
            'A'..='Z' | 'a'..='z' | '0'..='9' | '+' | '/' => {
                if eq_seen > 0 {
                    return false;
                }
            }
            '=' => {
                eq_seen += 1;
                if eq_seen > 2 {
                    return false;
                }
            }
            _ => return false,
        }
    }
    true
}
